File: src/leetcode/java/two_pointers/FourSum_18.java
package leetcode.java.two_pointers;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class FourSum_18 {
  public List<List<Integer>> fourSum(int[] nums, int target) {
    List<List<Integer>> result = new LinkedList<>();
    Arrays.sort(nums);
    for (int i = 0; i < nums.length - 3; i++) {
      if (nums[i] << 2 > target) break;
      if (i > 0 && nums[i] == nums[i - 1]) continue;
      for (int j = nums.length - 1; j > i + 2; j--) {
        if (nums[j] << 2 < target) break;
        if (j < nums.length - 1 && nums[j] == nums[j + 1]) continue;
        int lo = i + 1, hi = j - 1;
        int re = target - nums[i] - nums[j];
        int sum;
        while (lo < hi) {
          sum = nums[lo] + nums[hi];
          if (sum < re) lo++;
          else if (sum > re) hi--;
          else {
            result.add(Arrays.asList(nums[i], nums[lo], nums[hi], nums[j]));
            while (++lo < hi && nums[lo] == nums[lo - 1]);
            while (lo < --hi && nums[hi] == nums[hi + 1]);
          }
        }
      }
    }
    return result;
  }
}

File: src/leetcode/python/two_pointers/FourSum_18.py
class FourSum_18(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        result = []
        nums.sort()
        for i in range(len(nums) - 3):
            if nums[i] << 2 > target:
                break
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            for j in range(len(nums) - 1, i + 2, -1):
                if nums[j] << 2 < target:
                    break
                if j < len(nums) - 1 and nums[j] == nums[j + 1]:
                    continue
                lo, hi = i + 1, j - 1
                re = target - nums[i] - nums[j]
                while lo < hi:
                    subsum = nums[lo] + nums[hi]
                    if subsum > re:
                        hi -= 1
                    elif subsum < re:
                        lo += 1
                    else:
                        result.append([nums[i], nums[lo], nums[hi], nums[j]])
                        lo += 1
                        while lo < hi and nums[lo] == nums[lo - 1]:
                            lo += 1
                        hi -= 1
                        while lo < hi and nums[hi] == nums[hi + 1]:
                            hi -= 1

        return result
